src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,244 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <getopt.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include "sd-device.h"
+
+#include "alloc-util.h"
+#include "cgroup-util.h"
+#include "devnum-util.h"
+#include "device-util.h"
+#include "main-func.h"
+#include "path-util.h"
+#include "pretty-print.h"
+#include "verbs.h"
+
+static int help(void) {
+        printf("%s [OPTIONS...]\n\n"
+               "Set up iocost model and qos solutions for block devices\n"
+               "\nCommands:\n"
+               "  apply <path> [solution]    Apply the known solution for the device, if any, otherwise does nothing\n"
+               "  query <path>               Query the known solution for the device\n"
+               "\nOptions:\n"
+               "  -h --help                  Show this help\n"
+               "     --version               Show package version\n",
+               program_invocation_short_name);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+static int get_known_solutions(sd_device *device, char ***ret_solutions) {
+        _cleanup_free_ char **s = NULL;
+        const char *value;
+        int r;
+
+        assert(ret_solutions);
+
+        r = sd_device_get_property_value(device, "IOCOST_SOLUTIONS", &value);
+        if (r < 0)
+                return r;
+
+        s = strv_split(value, " ");
+        if (!s)
+                return -ENOMEM;
+
+        *ret_solutions = TAKE_PTR(s);
+
+        return 0;
+}
+
+static int query_named_solution(sd_device *device, const char *name, const char **model, const char **qos) {
```

Comment:
please follow coding style: please name return parameters "ret_xyz" and "ret_abc" and so on. mode/quos are return params, hence name them "ret_mode" and "ret_qos"